features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/metric/ViewEndedMetricDispatcherTest.kt

```diff
@@ -0,0 +1,280 @@
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.verifyNoMoreInteractions
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class)
+)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ForgeConfiguration(Configurator::class)
+internal class ViewEndedMetricDispatcherTest {
+
+    @Mock
+    lateinit var mockInternalLogger: InternalLogger
+
+    @FloatForgery(min = 0f, max = 1f)
+    private var fakeSampleRate: Float = 0f
+
+    @LongForgery(min = 0)
+    private var fakeDuration: Long = 0
+
+    @LongForgery(min = 0)
+    private var fakeLoadingTime: Long = 0
+
+    private lateinit var fakeViewType: RumViewType
+    private lateinit var fakeInvState: ViewInitializationMetricsState
+    private lateinit var fakeTnsState: ViewInitializationMetricsState
+
+    private lateinit var dispatcherUnderTest: ViewEndedMetricDispatcher
+
+    @BeforeEach
+    fun `set up`(forge: Forge) {
+        fakeViewType = forge.aValueFrom(RumViewType::class.java)
+        fakeTnsState = forge.aViewInitializationMetricsState(NoValueReason.TimeToNetworkSettle::class.java)
+        fakeInvState = forge.aViewInitializationMetricsState(NoValueReason.InteractionToNextView::class.java)
+
+        dispatcherUnderTest = ViewEndedMetricDispatcher(
+            viewType = fakeViewType,
+            internalLogger = mockInternalLogger,
+            samplingRate = fakeSampleRate
+        )
+    }
+
+    @Test
+    fun `M sendMetric only once W sendViewEnded`() {
+        // When
+        dispatcherUnderTest.sendViewEnded(fakeInvState, fakeTnsState)
+        // Then
+        verify(mockInternalLogger).logMetric(
+            messageBuilder = any(),
+            additionalProperties = any(),
+            samplingRate = eq(fakeSampleRate),
+            creationSampleRate = eq(null)
+        )
+
+        // When
+        dispatcherUnderTest.sendViewEnded(fakeInvState, fakeTnsState)
+
+        // Then
+        verify(mockInternalLogger).log(
+            eq(InternalLogger.Level.WARN),
+            target = eq(Target.TELEMETRY),
+            messageBuilder = any(),
+            throwable = eq(null),
+            onlyOnce = eq(false),
+            additionalProperties = eq(null)
+        )
+        verifyNoMoreInteractions(mockInternalLogger)
+    }
+
+    @Test
+    fun `M logMetric with expected arguments W sendViewEnded`() {
+        // Given
+        dispatcherUnderTest.onDurationResolved(fakeDuration)
+        dispatcherUnderTest.onViewLoadingTimeResolved(fakeLoadingTime)
+
+        // When
+        dispatcherUnderTest.sendViewEnded(fakeInvState, fakeTnsState)
+
+        // Then
+        verify(mockInternalLogger).logMetric(
+            messageBuilder = argThat { invoke() == VIEW_ENDED_MESSAGE },
+            additionalProperties = eq(expectedAttributes()),
+            samplingRate = eq(fakeSampleRate),
+            creationSampleRate = eq(null)
+        )
```

Comment:
to verify `InternalLogger` calls you can use `verifyLog` extension function, this will remove the need to call `invoke` for message assertion and also it won't be any need to use `eq`.